internal/rbd/rbd_util.go

```diff
@@ -564,6 +564,19 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
 
 		return err
 	}
+	defer func() {
+		// restore image from trash to avoid leaking stale image.
+		// If `trash remove` step fails, then in subsequent request, image should be
+		// found in the pool for deletion to resume normally. Hence, Restore the image
+		// from trash on err from `trash remove` step.
+		if err != nil {
```

Comment:
Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc

The way I was thinking to fix this one was

* If the image Not found in the pool do below operations as part of delete
* As we store the imageID check do we have any image in the trash with this ID
* In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash
* If the Image is found add a task to remove it from the trash
* If the image is not found return notFound or success as expected.

Reply:
> Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc

This mainly did focus on just errors from `trash remove` api calls.
 pod restart at a particular step is highly unlikely.

> 
> The way I was thinking to fix this one was
> 
> * If the image Not found in the pool do below operations as part of delete
> * As we store the imageID check do we have any image in the trash with this ID

We can directly restore image from trash here with imageID(if imageID != "") 
if found, send back an error and let another subsequent request handle it gracefully.
else, continue with cleanup.

> * In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash

IMO, we have been using and storing image ID for a very long time now and this is a cleanup bug patch and covering latest use cases is sufficient. 

The following steps increase complexity and add too many api calls.
(trash ls, find image, construct trash path, add task to remove  )

> * If the Image is found add a task to remove it from the trash
> * If the image is not found return notFound or success as expected.

@Madhu-1 , let me know what you think.

Reply:
> > Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc
> 
> This mainly did focus on just errors from `trash remove` api calls.
> pod restart at a particular step is highly unlikely.


Not sure why we want to fix it half not the full as we know the bug. IMO we should not work on only one scenario

> > The way I was thinking to fix this one was
> > 
> > * If the image Not found in the pool do below operations as part of delete
> > * As we store the imageID check do we have any image in the trash with this ID
> 
> We can directly restore image from trash here with imageID(if imageID != "")
> if found, send back an error and let another subsequent request handle it gracefully.
> else, continue with cleanup.

Why do you want to restore the image which is already in the trash (Have you verified how much time restore can take
depends on the image size? if it takes less time we can consider that but that is not enough to handle all the corner cases)


> > * In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash
> 
> IMO, we have been using and storing image ID for a very long time now and this is a cleanup bug patch and covering latest use cases is sufficient.
> 

Are we storing the imageID of the temp clone which we create during the PVC-PVC clone? and also consider the case where image creation got succeeded but next operation like storing the imageID in omap fails etc

> The following steps increase complexity and add too many api calls.
> (trash ls, find image, construct trash path, add task to remove )
> 

Am open to better suggestion to fix it completely not partially.
 
> > * If the Image is found add a task to remove it from the trash
> > * If the image is not found return notFound or success as expected.
> 
> @Madhu-1 , let me know what you think.